src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -30,6 +30,34 @@ internal void VerifyDiagnostics(string source, CSharpCompilationOptions options,
     [CompilerTrait(CompilerFeature.LocalFunctions)]
     public class LocalFunctionTests : LocalFunctionsTestBase
     {
+        [ConditionalFact(typeof(DesktopOnly))]
+        public void AwaitInLocalFuncInsideLock()
+        {
+            var source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class C
+{
+    void M(List<Task> listOfTasks)
+    {
+        lock(listOfTasks)
+        {
+            async Task localFunc()
+            {
+                await Task.Yield();
+            };
+
+            listOfTasks.Add(localFunc());
+        }
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics();
+        }
+
```

Comment:
The change should make rethrowing exceptions illegal in a local function in a catch, which is correct. I think we need a test for that.

Perhaps it is worth checking all the scenarios affected by the `AllClearedAtExecutableCodeBoundary`.  There are 6 flags in there. Not sure if `InCatchFilter` can be tested, since local function cannot be in an expression. Others are certainly testable. (you can put more than one scenario in a test, so this won't be a lot of tests)

Reply:
I just added some test scenarios. All flags but InCatchFilter are being tested.